src/lib.rs

```diff
@@ -1,3 +1,117 @@
+//!    byte of data. As other countries began to use computers, they created their own encodings to
+//!    include their non-latin letters (for example É in French). Asian languages had to widen the
+//!    character to 2 or more bytes since there are more than 256 glyphs in these languages.
+//!    Unicode was developed as a (successful as it turns out) attempt to unify all languages and
+//!    scripts into a single encoding scheme. Unicode characters are 4 bytes wide, meaning that
+//!    there is space for 4_294_967_296 different characters, which is expected to be enough. (It's
+//!    not quite this many in practice - some bit patterns are reserved).
+//!  - *utf-8* - Unicode was a very useful in making it easier to use the same software in
+//!    different parts of the world, but if all your text is latin then it takes 4 bytes per
+//!    character instead of 1 for ASCII. This would be quite wasteful, so the most popular way to
+//!    store unicode strings (utf-8) uses a continuation-bit style scheme to compress unicode
+//!    characters where possible. Another older scheme is utf-16, which works similarly, but uses 2
+//!    bytes (16 bits) as its smallest size.
+//!  - *grapheme, grapheme cluster* - Outside of computer typesetting, graphemes can have multiple
+//!    meanings in linguistics (see wikipedia). Within computer typesetting (and unicode in
+//!    particular), a grapheme cluster (called grapheme in the rest of these docs) can be thought
+//!    of as the conceptual version of a glyph. One exception to this is Han unification
+//!    (see below). It's important to note that a grapheme cluster may be 1 or more unicode
+//!    characters.  An example of a multi-character grapheme cluster is a rainbow flag (`U+1F3F3
+//!    (white flag), U+200D (zwj), U+1F308 (rainbow)`) which produces a single rainbow flag glyph.
+//!  - *script* - A set of graphemes that are used together to achieve written communcation.
+//!  - *language* - A context for deriving meaning from a script. Collections of graphemes from a
+//!    script will have different meanings in different languages, for example the word "pain"
+//!    means different things in English and French ("bread" in french).
+//!  - *locale* - Contextual information that defines how certain properties of the user interface
+//!    should look. Examples are whether text should be rendered left-to-right, or right-to-left
+//!    (as is the case in Hebrew, for example), and which varaint of Han characters to use.
+//!    Non-text locale information includes date format and timezone.
+//!  - *Han unification, CJK* - A number of Asian languages derive from ancient Chinese. The
+//!    acronym CJK expands to Chinese, Japanese, Korean. Traditional Taiwanese also has this
+//!    property, but since Taiwan has now adopted the Latin script it is less of an issue in
+//!    practice. These languages use different glyphs for the same unicode grapheme (the graphemes
```

Comment:
I would say "code point" rather than "grapheme." The word "grapheme" is seldom used in computer text outside the pleonasm "grapheme cluster."